libs/ml/include/ml/graph.hpp

```diff
@@ -96,6 +105,32 @@ class Graph
   std::vector<TrainablePtrType>                trainable_;
 };
 
+template <typename ArrayType>
```

Comment:
add doxygen style function comments and explanation of what each of the two set regularisation functions does differently